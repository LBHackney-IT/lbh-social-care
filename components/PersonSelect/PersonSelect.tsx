import React from 'react';
import s from './PersonSelect.module.scss';
import { LegacyResident } from 'types';
import { format } from 'date-fns';

interface ChoiceProps {
  value: number;
  name: string;
  label: string;
  hint: string;
  idToAdd: number | false;
  setIdToAdd: (value: number) => void;
}

const Choice = ({
  value,
  name,
  label,
  hint,
  idToAdd,
  setIdToAdd,
}: ChoiceProps): React.ReactElement => (
  <div className={`govuk-radios__item ${s.personResult}`}>
    <input
      className="govuk-radios__input"
      id={`${name}-${value}`}
      name={name}
      type="radio"
      value={value}
      aria-describedby={`${name}-${value}-hint`}
      checked={idToAdd === value}
      onChange={() => setIdToAdd(value)}
    />

    <label
      className={`govuk-radios__label lbh-body-s ${s.name}`}
      htmlFor={`${name}-${value}`}
    >
      {label}
    </label>

    <p id={`${name}-${value}-hint`} className={`lbh-body-xs ${s.meta}`}>
      {hint}
    </p>
  </div>
);

interface Props {
  label: string;
  people: LegacyResident[];
  idToAdd: number | false;
  setIdToAdd: (value: number) => void;
}

const PersonSelect = ({
  label,
  people,
  idToAdd,
  setIdToAdd,
}: Props): React.ReactElement => (
  <div className="govuk-form-group lbh-form-group">
    {console.log(people)}
    <fieldset className="govuk-fieldset">
      <legend className="govuk-label lbh-label">{label}</legend>
      <div
        className={`govuk-radios govuk-radios--small lbh-radios ${s.personList}`}
      >
        {people.map((person) => (
          <Choice
            name="person"
            label={`${person.firstName} ${person.lastName}`}
            value={person.mosaicId}
            idToAdd={idToAdd}
            setIdToAdd={setIdToAdd}
            key={person.mosaicId}
            hint={`#${person.mosaicId} · Born ${format(
              new Date(String(person.dateOfBirth)),
              'd MMM yyyy'
            )} · ${person.address?.address}`}
            {...people}
          />
        ))}
      </div>
    </fieldset>
  </div>
);

export default PersonSelect;
