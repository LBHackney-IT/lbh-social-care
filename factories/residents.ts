import { Factory } from 'fishery';

import { LegacyResident, Resident } from 'types';

export const legacyResidentFactory = Factory.define<LegacyResident>(
  ({ sequence }) => ({
    mosaicId: String(sequence),
    dateOfBirth: '2020-11-13',
    firstName: 'Foo',
    lastName: 'Bar',
    nhsNumber: '12345',
    ageContext: 'A',
    gender: 'F',
    address: {
      address: 'sjakdjlk',
      postcode: 'hdsadjk',
    },
  })
);

export const residentFactory = Factory.define<Resident>(({ sequence }) => ({
  id: sequence,
  dateOfBirth: '2020-11-13',
  firstName: 'Foo',
  lastName: 'Bar',
  nhsNumber: 12345,
  contextFlag: 'A',
  gender: 'F',
  createdBy: 'foo@bar.com',
  otherNames: [],
  phoneNumbers: [],
  addresses: [],
  address: {
    address: 'sjakdjlk',
    postcode: 'hdsadjk',
  },
}));

export const mockedLegacyResident = legacyResidentFactory.build();
export const mockedResident = residentFactory.build();
