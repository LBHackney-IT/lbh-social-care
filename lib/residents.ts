import axios from 'axios';

import type { LegacyResident, Resident, ResidentsAPI, AgeContext } from 'types';

const { ENDPOINT_API, AWS_KEY } = process.env;

const headers = { 'x-api-key': AWS_KEY };

interface ResidentBE extends Omit<LegacyResident, 'restricted'> {
  restricted: 'Y' | 'N';
  addressList: Array<{
    displayAddressFlag: 'Y' | 'N';
    addressLine1: string;
    postCode: string;
    uprn: string;
  }>;
}

const sanitiseResidentData = (residents: ResidentBE[]): LegacyResident[] =>
  residents?.map(
    ({ addressList, ...resident }: ResidentBE): LegacyResident => {
      const address = addressList?.find(
        ({ displayAddressFlag }) => displayAddressFlag === 'Y'
      );
      return {
        ...resident,
        restricted: resident.restricted === 'Y',
        address: address && {
          address: address.addressLine1,
          postcode: address.postCode,
          uprn: address.uprn,
        },
      };
    }
  );

export const getResidents = async (
  params: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers,
    params,
  });
  return { ...data, residents: sanitiseResidentData(data.residents) };
};

export const getResident = async (
  personId: number,
  params: { context_flag: AgeContext }
): Promise<Resident> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents/${personId}`, {
    headers,
    params,
  });
  return { ...data, restricted: data.restricted === 'Y' };
};

export const normalisePhoneInput = (formData: {
  phoneNumbers?: Record<string, string>[];
  [key: string]: unknown;
}): Record<string, unknown> => ({
  ...formData,
  phoneNumbers: formData.phoneNumbers
    ? formData.phoneNumbers.map((phone) => ({
        ...phone,
        type: phone.type === '' ? 'main' : phone.type,
      }))
    : null,
});

export const addResident = async (
  formData: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.post(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};

export const updateResident = async (
  formData: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.patch(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};
