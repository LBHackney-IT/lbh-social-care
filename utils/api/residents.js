import axios from 'axios';

export const getResidents = async (params) => {
  const { data } = await axios.get('/api/residents', {
    params,
  });
  return data;
};

export const getResident = async (id, params) => {
  const { data } = await axios.get(`/api/residents/${id}`, {
    params,
  });
  return data;
};

export const addResident = async (formData) => {
  const { data } = await axios.post(`/api/residents`, formData);
  return data;
};
