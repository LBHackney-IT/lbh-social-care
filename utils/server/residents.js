import axios from 'axios';

const {
  ENDPOINT_API,
  ENDPOINT_MOSAIC,
  AWS_KEY,
  AWS_AUTHORIZATION,
} = process.env;

export const getResidents = async (params) => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers: {
      'x-api-key': AWS_KEY,
    },
    params,
  });
  return data.residents;
};

export const getResident = async (id) => {
  const { data } = await axios.get(`${ENDPOINT_MOSAIC}/residents/${id}`, {
    headers: {
      Authorization: AWS_AUTHORIZATION,
    },
  });
  return data;
};
