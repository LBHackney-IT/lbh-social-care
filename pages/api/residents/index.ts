import { StatusCodes } from 'http-status-codes';

import { getResidents, addResident } from 'lib/residents';
import { isAuthorised } from 'utils/auth';

import type { NextApiRequest, NextApiResponse, NextApiHandler } from 'next';

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const user = isAuthorised(req);
  if (!user) {
    return res.status(StatusCodes.UNAUTHORIZED).end();
  }
  if (!user.isAuthorised) {
    return res.status(StatusCodes.FORBIDDEN).end();
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getResidents({
          ...req.query,
          context_flag: user.permissionFlag,
        });
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error('Residents get error:', error?.response?.data);
        error?.response?.status === StatusCodes.NOT_FOUND
          ? res
              .status(StatusCodes.NOT_FOUND)
              .json({ message: 'Residents Not Found' })
          : res
              .status(StatusCodes.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Residents' });
      }
      break;

    case 'POST':
      try {
        const data = await addResident(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error('Resident post error:', error?.response?.data);
        console.error('Resident post request:', req);
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add resident' });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default endpoint;
