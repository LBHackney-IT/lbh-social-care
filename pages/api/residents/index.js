import * as HttpStatus from 'http-status-codes';

import { getResidents, addResident } from 'utils/server/residents';
import { isAuthorised } from 'utils/auth';

export default async (req, res) => {
  if (!isAuthorised(req)) {
    return res
      .status(HttpStatus.UNAUTHORIZED)
      .json({ message: 'Auth cookie missing.' });
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getResidents(req.query);
        res.status(HttpStatus.OK).json(data);
      } catch (error) {
        console.log('Residents get error:', error?.response?.data);
        error?.response?.status === HttpStatus.NOT_FOUND
          ? res
              .status(HttpStatus.NOT_FOUND)
              .json({ message: 'Residents Not Found' })
          : res
              .status(HttpStatus.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Residents' });
      }
      break;

    case 'POST':
      try {
        const data = await addResident(req.body);
        res.status(HttpStatus.OK).json(data);
      } catch (error) {
        console.log('Resident post error:', error?.response?.data);
        res
          .status(HttpStatus.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add resident' });
      }
      break;

    default:
      res
        .status(HttpStatus.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};
