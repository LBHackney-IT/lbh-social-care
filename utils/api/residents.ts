import axios from 'axios';
import useSWR, {
  useSWRInfinite,
  SWRResponse,
  SWRInfiniteSWRResponse,
} from 'swr';

import { getInfiniteKey } from 'utils/api';

import type { Resident, ErrorAPI } from 'types';

export const useResidents = (
  params: Record<string, unknown>,
  invoke = true
): SWRInfiniteSWRResponse<Record<string, unknown>, Error> =>
  useSWRInfinite(
    // @ts-ignore
    invoke ? getInfiniteKey('/api/residents', 'residents', params) : null
  );

export const useResident = (
  personId: number
): SWRResponse<Resident, ErrorAPI> => useSWR(`/api/residents/${personId}`);

export const addResident = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.post(`/api/residents`, formData);
  return { ref: data?.personId, data };
};

export const updateResident = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/residents`, formData);
  return { ref: data?.personId, data };
};
