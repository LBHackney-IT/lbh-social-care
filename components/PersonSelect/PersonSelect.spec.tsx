import PersonSelect from './PersonSelect';
import { render, screen } from '@testing-library/react';
import { mockedResident } from 'factories/residents';

describe('PersonSelect', () => {
  it("correctly formats a person's details", () => {
    render(
      <PersonSelect
        name="person"
        label="Matching people"
        people={[mockedResident]}
      />
    );
    expect(screen.getByText('foo bar'));
    expect(screen.getByText('Born 13 Nov 2020 Â· Example address'));
    expect(screen.getAllByRole('radio').length).toBe(1);
  });

  //   TODO: handle incomplete details?

  //   TODO: Actually manage and submit state, once we have reason for that sort of thing
});
