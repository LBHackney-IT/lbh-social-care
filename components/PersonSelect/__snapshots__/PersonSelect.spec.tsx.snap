// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonSelect should render properly 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group lbh-form-group"
  >
    <fieldset
      class="govuk-fieldset"
    >
      <legend
        class="govuk-label lbh-label"
      >
        Matching people
      </legend>
      <div
        class="govuk-radios govuk-radios--small lbh-radios personList"
      >
        <div
          class="govuk-radios__item personResult"
        >
          <input
            aria-describedby="person-1-hint"
            checked=""
            class="govuk-radios__input"
            id="person-1"
            name="person"
            type="radio"
            value="1"
          />
          <label
            class="govuk-radios__label lbh-body-s name"
            for="person-1"
          >
            Foo Bar
          </label>
          <p
            class="lbh-body-xs meta"
            id="person-1-hint"
          >
            #1 · Born 13 Nov 2020 · sjakdjlk
          </p>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;
